File: utils/aalmodel.py
import copy
import types
import time
import traceback
import fmbt

SILENCE = -3

def setCodeFileLine(c, filename, lineno, funcname=None):
    if funcname == None:
        funcname = c.co_name
    return types.CodeType(
        c.co_argcount, c.co_nlocals, c.co_stacksize, c.co_flags,
        c.co_code, c.co_consts, c.co_names, c.co_varnames,
        filename, funcname, lineno, c.co_lnotab, c.co_freevars)

class AALModel:
    def __init__(self, model_globals):
        self._all_guards = self._get_all("guard", "action")
        self._all_bodies = self._get_all("body", "action")
        self._all_adapters = self._get_all("adapter", "action")
        self._all_names = self._get_all("name", "action")
        self._all_types = self._get_all("type", "action")
        self._all_tagnames = self._get_all("name", "tag")
        self._all_tagguards = self._get_all("guard", "tag")
        self._all_tagadapters = self._get_all("adapter", "tag")
        if len(self._get_all("guard_next_block", "serial")) > 0:
            self._has_serial = True
        else:
            self._has_serial = False
        self._variables = model_globals
        self._variables['action'] = lambda name: self._all_names.index(name) + 1
        self._variables['name'] = lambda name: self._all_names.index(name)
        self._variables['variable'] = lambda varname: self._variables[varname]
        self._variables['assign'] = lambda varname, v: self._variables.__setitem__(varname, v)
        self._variables['guard_list'] = []
        self._stack = []
        self._stack_executed_actions = []
        self._adapter_exit_executed = False
        self._enabled_actions_stack = [set()]
        fmbt._g_testStep = 0

    def _get_all(self, property_name, itemtype):
        plist = []
        i = 1
        while 1:
            try: plist.append(getattr(self, itemtype + str(i) + property_name))
            except: return plist
            i += 1

    def _set_all_class(self, property_name, itemtype, value_array):
        i = 1
        cls = self.__class__
        while 1:
            itemname = itemtype + str(i) + property_name
            if hasattr(cls, itemname):
                setattr(cls, itemname, value_array[i-1])
            else:
                return
            i += 1

    def call(self, func, call_arguments = ()):
        guard_list = None
        try:
            func_name = func.__name__
            if func_name.endswith("guard"):
                guard_list = self._variables['guard_list']
                guard_list.append(
                    getattr(self, func_name.replace("guard", "name"), ""))
            fmbt._g_simulated_actions = self._stack_executed_actions
            if hasattr(func,"requires"):
                for prerequire in func.requires:
                    if not self.call(getattr(self,prerequire)):
                        return False
            args = []
            for arg in call_arguments:
                if arg == '':
                    args.append('')
                else:
                    args.append(eval(arg, self._variables))
            self._variables['args'] = args
            return eval(func.func_code, self._variables)
        except Exception, e:
            self._log("Exception %s in %s: %s" % (e.__class__, func.func_name, e))
            self._log(traceback.format_exc())
            raise
        finally:
            if guard_list != None:
                guard_list.pop()

    def call_exception_handler(self, handler_name, action_name, exc):
        rv = self._variables[handler_name](action_name, exc)
        if type(rv) == int:
            return rv
        elif rv == None or rv == True:
            return self._variables['action'](action_name)
        else:
            raise Exception('''Exception handler "%s('%s', %s)" returned unexpected value: %s''' %
                            (handler_name, action_name, exc, rv))

    def reset(self):
        # initialize model
        fmbt._g_actionName = "AAL: initial_state"
        rv = self.call(self.initial_state)
        self._push_variables = [
            v for v in self.__class__.push_variables_set
            if (v in self._variables and
                type(eval(v, self._variables)) not in [types.ModuleType, types.ClassType])
            ]
        return rv

    def adapter_init():
        return True

    def init(self):
        # initialize adapter
        fmbt._g_actionName = "AAL: adapter_init"
        rv = self.call(self.adapter_init)
        return rv

    def adapter_exit(verdict, reason):
        return

    def aexit(self, verdict, reason):
        if not self._adapter_exit_executed:
            self._adapter_exit_executed = True
            fmbt._g_actionName = "AAL: adapter_exit"
            self.adapter_exit.im_func(verdict, reason)

    def adapter_execute(self, i, adapter_call_arguments = ()):
        if not 0 < i <= len(self._all_names):
            raise IndexError('Cannot execute action %s adapter code' % (i,))
        if self._all_types[i-1] == "input":
            try:
                fmbt._g_actionName = self._all_names[i-1]
                fmbt._g_testStep += 1
                rv = self.call(self._all_adapters[i-1], adapter_call_arguments)
                fmbt._g_testStep -= 1
                if rv == None: return i
                else: return rv
            except Exception, exc:
                if 'adapter_exception_handler' in self._variables:
                    return self.call_exception_handler('adapter_exception_handler', self._all_names[i-1], exc)
                else:
                    raise
        else:
            self._log("AAL model: adapter_execute for an output action in AAL." +
                      "This should take place in observe().\n")
            return 0

    def tag_execute(self, i):
        if not 0 < i <= len(self._all_tagnames):
            raise IndexError('Cannot execute tag %s adapter code' % (i,))
        fmbt._g_actionName = "tag: " + self._all_tagnames[i-1]
        rv = self.call(self._all_tagadapters[i-1])
        return rv

    def model_execute(self, i):
        if not 0 < i <= len(self._all_names):
            # If adapter execute returns 0, that is reports unidentified action,
            # test engine checks if executing an unidentified action is ok by
            # calling model_execute(0). In AAL/Python it is never ok.
            return 0
        fmbt._g_actionName = self._all_names[i-1]
        if i in self._enabled_actions_stack[-1] or self.call(self._all_guards[i-1]):
            self.call(self._all_bodies[i-1])
            if self._has_serial:
                for postfunc in getattr(self, self._all_bodies[i-1].__name__ + "_postcall", []):
                    getattr(self, postfunc)(fmbt._g_actionName)
            if len(self._stack) == 0:
                fmbt._g_testStep += 1
            if len(self._stack_executed_actions) > 0:
                self._stack_executed_actions[-1].append(fmbt._g_actionName)
            self._enabled_actions_stack[-1] = set()
            return i
        else:
            self._enabled_actions_stack[-1] = set()
            return 0

    def getActions(self):
        enabled_actions = []
        for index, guard in enumerate(self._all_guards):
            fmbt._g_actionName = self._all_names[index]
            if self.call(guard): enabled_actions.append(index + 1)
        self._enabled_actions_stack[-1] = set(enabled_actions)
        return enabled_actions

    def getIActions(self):
        enabled_iactions = []
        try:
            for index, guard in enumerate(self._all_guards):
                fmbt._g_actionName = self._all_names[index]
                if self._all_types[index] == "input" and self.call(guard):
                    enabled_iactions.append(index + 1)
        except Exception, e:
            raise Exception('Error at guard() of "%s": %s: %s' % (
                self._all_names[index], type(e).__name__, e))
        self._enabled_actions_stack[-1].update(enabled_iactions)
        return enabled_iactions

    def getprops(self):
        enabled_tags = []
        for index, guard in enumerate(self._all_tagguards):
            fmbt._g_actionName = "tag: " + self._all_tagnames[index]
            if self.call(guard): enabled_tags.append(index + 1)
        return enabled_tags

    def getActionNames(self):
        return self._all_names

    def getSPNames(self):
        return self._all_tagnames

    def push(self):
        # initial state must reset all variables.
        # automatic push saves only their states
        stack_element = {}
        for varname in self._push_variables:
            stack_element[varname] = copy.deepcopy(self._variables[varname])
        if self._has_serial:
            stack_element["!serial_abn"] = copy.deepcopy(self._get_all("guard_next_block", "serial"))
        self._stack.append(stack_element)
        self._stack_executed_actions.append([])
        self._enabled_actions_stack.append(set(self._enabled_actions_stack[-1]))

    def pop(self):
        stack_element = self._stack.pop()
        self._stack_executed_actions.pop()
        for varname in stack_element:
            if varname.startswith("!"): continue
            self._variables[varname] = stack_element[varname]
        if self._has_serial:
            self._set_all_class("guard_next_block", "serial", stack_element["!serial_abn"])
        self._enabled_actions_stack.pop()

    def state(self, discard_variables = set([]), include_variables=None):
        """
        Return the current state of the model as a string.
        By comparing strings one can check if the state is already seen.
        """
        rv_list = []
        for varname in self._push_variables:
            if ((include_variables and not varname in include_variables) or
                (varname in discard_variables)):
                continue
            rv_list.append("%s = %s" % (varname, repr(self._variables[varname])))
        if self._has_serial:
            rv_list.append("!serial = %s" % (self._get_all("guard_next_block", "serial"),))
        return '\n'.join(rv_list)

    def observe(self, block):
        poll_more = True
        start_time = 0

        # Executing adapter blocks of output actions is allowed to
        # change the state of the model. Allow execution of outputs
        # whose guards are true both before executing adapter blocks*
        # or after it. For that purpose, add currently enabled output
        # actions to enabled_actions_stack.
        enabled_oactions = []
        for index, guard in enumerate(self._all_guards):
            fmbt._g_actionName = self._all_names[index]
            if (self._all_types[index] == "output" and
                not (index + 1) in self._enabled_actions_stack[-1] and
                self.call(guard)):
                enabled_oactions.append(index + 1)
        self._enabled_actions_stack[-1].update(enabled_oactions)

        while poll_more:
            for index, adapter in enumerate(self._all_adapters):
                if self._all_types[index] != "output": continue
                fmbt._g_actionName = self._all_names[index]
                output_action = self.call(adapter)
                observed_action = None
                if type(output_action) == str:
                    observed_action = self._all_names.index(output_action) + 1
                elif type(output_action) == type(True) and output_action == True:
                    observed_action = index + 1
                elif type(output_action) == int and output_action > 0:
                    observed_action = output_action

                if observed_action:
                    self._log('observe: action "%s" adapter() returned %s. Reporting "%s"' % (
                            self._all_names[index], output_action,
                            self._all_names[observed_action-1]))
                    return [observed_action]
            if block:
                if not start_time: start_time = time.time()
                elif time.time() - start_time > self.timeout:
                    return [SILENCE]
            else:
                poll_more = False
        return [SILENCE]

File: utils/aalmodel.rb
class AALModel
    attr_accessor :log 
    attr_accessor :timeout
    @@SILENCE = -3

    def initialize()
        @immutable_types = Set.new(["String", "Fixnum", "Long", "Float", "TrueClass", "NilClass", "Array"])

        @variables = {}
        @variables['guard_list'] = []

        @aal_block_name = {}
        @all_guards = get_all("guard", "action")
        @all_bodies = get_all("body", "action")
        @all_adapters = get_all("adapter", "action")
        @all_names = get_all("name", "action")
        @all_types = get_all("type","action")

        @all_tagnames = get_all("name", "tag")
        @all_tagguards = get_all("guard", "tag")
        @all_tagadapters = get_all("adapter", "tag")
        
        if get_all("guard_next_block", "serial").length > 0
            @has_serial = true
        else
            @has_serial = false
        end

        @stack = []
        @enabled_actions_stack = [Set.new()]
        @stack_executed_actions = []
        @push_variables = []
    end
    def call(func_name, call_arguments=nil)
        guard_list = nil
        begin
            Fmbt.simulated_actions = @stack_executed_actions
            if func_name[-1,1] == "d" # faster than func_name.endswith("guard")
                guard_list = @variables['guard_list']
                guard_list.push(@aal_block_name[func_name])
                #todo  dont konw what is it
                # for prerequisite in func.requires
                    # if not self.call(getattr(self, prerequisite))
                    #     return False
                    # end
                # end
            end
            if call_arguments
                args = []
                for arg in call_arguments
                    if arg == ''
                        args.push('')
                    else
                        args.push(eval(arg, self._variables))
                    end
                end
                @variables['args'] = args
            else
                @variables['args'] = []
            end
            rv = self.send(func_name)
            if guard_list
                guard_list.pop()
            end
            return rv
        rescue Exception, e
            @log.log("Exception #{e.class} in #{func.func_name} #{e.message}")
            @log.log(traceback.format_exc())
            if guard_list
                guard_list.pop()
            raise
            end
        end
    end

    def get_all(property_name, itemtype)
        plist = []
        i = 1
        while true
            obj = self.instance_variable_get("@#{itemtype}#{i}#{property_name}")
            if obj != nil
                plist.push(obj)
            else 
                #methos in ruby 1.8.6 returns string
                obj = self.methods.include?("#{itemtype}#{i}#{property_name}")
                #methods in ruby 2.3 returns symbol
                obj_symbol_check = self.methods.include?("#{itemtype}#{i}#{property_name}".to_sym)
                if obj or obj_symbol_check
                    if ["guard", "body", "adapter"].include?(property_name) and ["action", "tag"].include?(itemtype)
                        @aal_block_name["#{itemtype}#{i}#{property_name}"] = self.instance_variable_get("@#{itemtype}#{i}name")
                    elsif itemtype == "serial"
                        @aal_block_name["serial#{i}#{property_name}"] = self.instance_variable_get("@serial#{i}name")
                    end
                    plist.push("#{itemtype}#{i}#{property_name}")
                else
                    return plist
                end
            end
            i += 1
        end
    end

    def getActions()
        enabled_actions = []
        @all_guards.each_with_index do |guard,index|
            Fmbt.actionName = @all_names[index]
            if call(guard)
                enabled_actions.push(index + 1)
            end
        end
        @enabled_actions_stack[-1] = Set.new(enabled_actions)
        return enabled_actions
    end
    
    def getActionNames()
        return @all_names
    end

    def getSPNames()
        return @all_tagnames
    end

    def push(obj=nil)
        """
        Push current state (or optional parameter obj) to the stack.

        See also pop(), state_obj().
        """
        if obj != nil
            @stack.push(obj[0])
            @stack_executed_actions.push(obj[1])
            @enabled_actions_stack.push(obj[2])
            return
        end
        # initial state must reset all variables.
        # automatic push saves only their states
        stack_element = {}
        for varname in @push_variables
            val = @variables[varname]
            if @immutable_types.include?(val.class)
                stack_element[varname] = val
            else
                #todo  deep copy
                stack_element[varname] = val
            end 
        end
        if @has_serial
            #todo  deep copy
            stack_element["!serial_abn"] = get_all("guard_next_block", "serial")
        end
        @stack.push(stack_element)
        @stack_executed_actions.push([])
        @enabled_actions_stack.push(Set.new(@enabled_actions_stack[-1]))
    end

    def pop()
        """
        Pop the topmost state in the stack and start using it as the
        current state.

        See also push()
        """
        stack_element = @stack.pop()
        @stack_executed_actions.pop()
        for varname in stack_element
            if varname[0..1] == "!"
                 next 
            end
            @variables[varname] = stack_element[varname]
        end
        if @has_serial
            set_all_class("guard_next_block", "serial", stack_element["!serial_abn"])
        end
        @enabled_actions_stack.pop() 
    end
    
    def stack_top()
        return @stack[-1], @stack_executed_actions[-1], @enabled_actions_stack[-1]
    end
    
    def stack_discard()
        @stack.pop()
        @stack_executed_actions.pop()
        @enabled_actions_stack.pop()
    end

    def state_obj()
        """
        Return current state as a Python object
        """
        push()
        obj = stack_top()
        stack_discard()
        return obj
    end

    def set_all_class(property_name, itemtype, value_array)
        i = 1
        cls = self.class
        while true
            itemname = itemtype + i.to_s + property_name
            if self.instance_variable.include?(itemname)
                self.instance_variable_set(itemname, value_array[i-1])
            else
                return
            end
            i += 1
        end 
    end

    def set_state_obj(obj):
        """
        Set obj as current state. See also state_obj().
        """
        push(obj)
        pop() 
    end

    def state_obj_copy(obj=nil):
        """
        Return copy of a state_obj.
        Faster than copy.deepcopy(self.state_obj())
        """
        if obj == nil:
            obj = state_obj()
        end
        stack_top, stack_executed_top, stack_enabled_top = obj
        copy_stack_top = {}
        for varname in @push_variables
            val = stack_top[varname]
            if @immutable_types.include?(val.class)
                copy_stack_top[varname] = val
            else
                #todo : deepcopy
                copy_stack_top[varname] = copy.deepcopy(val)
            end
        end
        if @has_serial:
            #todo : ddeepcopy
            copy_stack_top["!serial_abn"] = stack_top["!serial_abn"]
        end
        copy_stack_executed_top = list(stack_executed_top)
        copy_stack_enabled_top = Set.new(stack_enabled_top)
        return copy_stack_top, copy_stack_executed_top, copy_stack_enabled_top
    end 
    
    def reset()
        # initialize model
        Fmbt.actionName = "AAL initial_state"
        rv = call('initial_state')
        for v in @push_variables_set
            @push_variables.push(v)
        end
        return rv
    end

    def init()
        # initialize adapter
        Fmbt.actionName = "AAL adapter_init"
        rv = call('adapter_init')
        return rv
    end
    
    #getting enabled tags
    def getprops()
        enabled_tags = []
        @all_tagguards = get_all("guard", "tag")
        @all_tagguards.each_with_index do |guard,index|
            Fmbt.actionName = "tag " + @all_tagnames[index]
            if call(guard)
                enabled_tags.push(index + 1)
            end
        end
        return enabled_tags
    end

    def tag_execute(i)
        if not (i > 0 and i <= @all_tagnames.length)
            raise IndexError.new("Cannot execute tag #{i} adapter code")
        end
        Fmbt.actionName = "tag " + @all_tagnames[i-1]
        begin
            rv = call(@all_tagadapters[i-1])
        rescue Exception => e
            if  @variables.include?('adapter_exception_handler')
                return call_tagexception_handler('adapter_exception_handler', @all_tagnames[i-1], exc)
            else
                raise
            end
        end
        return rv
    end

    def state(discard_variables=nil, include_variables=nil)
        """
        Return the current state of the model as a string.
        By comparing strings one can check if the state is already seen.
        """
        rv_list = []
        for varname in @push_variables
            if (include_variables and not include_variables.include?(varname)) or (discard_variables and discard_variables.include?(varname))
                next
            end
            rv_list.push("#{varname} = '#{eval("$#{varname}")}'")
        end
        if @has_serial
            rv_list.push("!serial = #{get_all("guard_next_block", "serial")}")
        end
        return rv_list.join('\n')
    end
       
    def observe(block)
        poll_more = true
        start_time = 0
        
        # Executing adapter blocks of output actions is allowed to
        # change the state of the model. Allow execution of outputs
        # whose guards are true both before executing adapter blocks*
        # or after it. For that purpose, add currently enabled output
        # actions to enabled_actions_stack.
        enabled_oactions = []
        @all_guards.each_with_index do |guard,index|
            Fmbt.actionName = @all_names[index]
            if @all_types[index] == "output" and not @enabled_actions_stack[-1].include?(index + 1)  and call(guard)
                enabled_oactions.append(index + 1)
            end
        end
        @enabled_actions_stack[-1].merge(enabled_oactions)

        while poll_more
            @all_adapters.each_with_index do |adapter,index|
                if @all_types[index] != "output"
                    next
                end
                Fmbt.actionName = @all_names[index]
                begin
                    fmbt._g_testStep += 1
                    output_action = self.call(adapter)
                rescue Exception => e
                    if @variables.include?('adapter_exception_handler')
                        output_action = call_exception_handler(
                            'adapter_exception_handler',
                            @all_names[index], exc,
                            pass_through_rv = [False])
                    else
                        raise
                    end
                ensure
                    Fmbt.testStep -= 1
                end
                observed_action = None
                if type(output_action) == str
                    observed_action = @all_names.index(output_action) + 1
                elsif type(output_action) == type(True) and output_action
                    observed_action = index + 1
                elsif type(output_action) == int and output_action > 0
                    observed_action = output_action
                end
                if observed_action
                    log.log("observe action \"#{@all_names[index]}\" adapter() returned #{output_action}. Reporting \"#{@all_names[observed_action-1]}\"")
                    Fmbt.lastExecutedActionName = @all_names[observed_action-1]
                    return [observed_action]
                end
            end
            if block
                if not start_time
                     start_time = time.time()
                elsif time.time() - start_time > timeout
                    return [@@SILENCE]
                end
            else
                poll_more = false
            end
        end
        return [@@SILENCE]
    end
     
    
end